lib/article_json/import/google_doc/html/parser.rb

```diff
@@ -0,0 +1,109 @@
+module ArticleJSON
+  module Import
+    module GoogleDoc
+      module HTML
+        class Parser
+          # @param [String] html
+          def initialize(html)
+            doc = Nokogiri::HTML(html)
+            @body_enumerator = doc.xpath('//body').last.children.to_enum
+
+            css_node = doc.xpath('//head/style').last
+            @css_analyzer = CSSAnalyzer.new(css_node&.inner_text)
+          end
+
+          # Parse the body of the document and return the result
+          # @return [Array]
+          def parsed_content
+            @parsed_content ||= parse_body
+          end
+
+          private
+
+          # Loop over all body nodes and parse them
+          # @return [Array]
+          def parse_body
+            @parsed_content = []
+            while body_has_more_nodes?
+              @parsed_content << begin
+                @current_node = NodeAnalyzer.new(@body_enumerator.next)
+                parse_current_node || next
+              end
+            end
+            @parsed_content
+          end
+
+          # Parse the current node and return an element, if available
+          # @return [Object]
+          def parse_current_node
+            case @current_node.type
+            when :heading then parse_heading
+            when :paragraph then parse_paragraph
+            when :list then parse_list
+            when :image then parse_image
+            when :text_box then parse_text_box
+            when :quote then parse_quote
+            when :hr, :empty, :unknown then nil
+            end
```

Comment:
Not to be critic, and not sure if that would be really rubyist, but I feel implementing the whole HTML standards might be a very long case. Might be interesting to consider something like a hash of kinds of node to functions? Maybe something different but cleaner than just a hash, for example, a hookable code architecture?
These are not strong suggestions, just ideas coming raw to my mind.

Reply:
It's not really HTML, it's the elements we are supporting (text box for example is something we made up)...

I thought about doing this via magic methods or a hash, but found this easiest to read and understand at a first glance...

Reply:
I agree. There probably should be a registry of classes for the different elements which then have their own parse method and maybe detection too. 
Case statements tend to be a bad smell.

Reply:
I actually intended to implement the parsing in the individual classes.
But the problem is that some methods need to move the "cursor" (on the enumerator) until they find an `<hr>`  (like the one for quotes or text boxes).
Having the handling of the enumerator distributed over several classes is probably not a good idea.

I can refactor the case statement into a registry-hash though...